projects/plugins/jetpack/_inc/lib/core-api/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-top-posts.php

```diff
@@ -0,0 +1,206 @@
+								return is_string( $param );
+							},
+						),
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Get the site's post types.
+	 *
+	 * @return array Site's post types.
+	 */
+	public function get_post_types() {
+		$post_types       = array_values( get_post_types( array( 'public' => true ) ) );
+		$post_types_array = array();
+
+		foreach ( $post_types as $type ) {
+			$post_types_array[] = array(
+				'label' => get_post_type_object( $type )->labels->name,
+				'id'    => $type,
+			);
+		}
+
+		return $post_types_array;
+	}
+
+	/**
+	 * Get the site's top content.
+	 *
+	 * @param \WP_REST_Request $request request object.
+	 *
+	 * @return array Data on top posts.
+	 */
+	public function get_top_posts( $request ) {
+		$period        = $request->get_param( 'period' );
+		$all_time_days = floor( ( time() - strtotime( get_option( 'site_created_date' ) ) ) / ( 60 * 60 * 24 * 365 ) );
+
+		// While we only display ten posts, users can filter out content types.
+		// As such, we should obtain a few spare posts from the Stats endpoint.
+		$posts_to_obtain_count = 30;
+
+		// We should not override cache when displaying the block on the frontend.
+		// But we should allow instant preview of changes when editing the block.
+		$is_rendering_block = isset( $request['types'] );
+		$override_cache     = ! $is_rendering_block;
+
+		$query_args = array(
+			'max'       => $posts_to_obtain_count,
+			'summarize' => true,
+			'num'       => $period !== 'all-time' ? $period : $all_time_days,
+			'period'    => 'day',
+		);
+
+		$data = ( new WPCOM_Stats() )->get_top_posts( $query_args, $override_cache );
+
+		if ( is_wp_error( $data ) ) {
+			$data = array( 'summary' => array( 'postviews' => array() ) );
+		}
+
+		$posts_retrieved = count( $data['summary']['postviews'] );
+
+		// Fallback to random posts if user does not have enough top content.
+		if ( $posts_retrieved < $posts_to_obtain_count ) {
+			$args = array(
+				'numberposts' => $posts_to_obtain_count - $posts_retrieved,
+				'exclude'     => array_column( $data['summary']['postviews'], 'id' ),
+				'orderby'     => 'rand',
+				'post_status' => 'publish',
+			);
+
+			$random_posts = get_posts( $args );
+
+			foreach ( $random_posts as $post ) {
+				$random_posts_data = array(
+					'id'     => $post->ID,
+					'href'   => get_permalink( $post->ID ),
+					'date'   => $post->post_date,
+					'title'  => $post->post_title,
+					'type'   => 'post',
+					'public' => true,
+				);
+
+				$data['summary']['postviews'][] = $random_posts_data;
+			}
+
+			$data['summary']['postviews'] = array_slice( $data['summary']['postviews'], 0, 10 );
+		}
+
+		$top_posts = array();
+
+		foreach ( $data['summary']['postviews'] as $post ) {
+			$post_id   = $post['id'];
+			$thumbnail = get_the_post_thumbnail_url( $post_id );
+
+			if ( ! $thumbnail ) {
+				$post_image = reset( get_attached_media( 'image', $post_id ) );
+				$thumbnail  = wp_get_attachment_url( $post_image->ID );
```

Comment:
```suggestion
				$post_images = get_attached_media( 'image', $post_id );
				$post_image = reset( $post_images );
				if ( $post_image ) {
					$thumbnail  = wp_get_attachment_url( $post_image->ID );
				}
```

Discovered two warning here:

1. reset requires a variable
2. reset could return false if the array is empty